Add support for playwright e2e tests

c0d33ngr opened the pull request:
PR to close issue #1650

c0d33ngr committed:
Add some playwright-e2e tests conversion

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/jhipster/generator-jhipster-svelte?pullRequest=1954) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/jhipster/generator-jhipster-svelte?pullRequest=1954) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/jhipster/generator-jhipster-svelte?pullRequest=1954) it.</sub>

c0d33ngr committed:
Add some test cases

c0d33ngr committed:
Add some test cases

c0d33ngr committed:
Add more test conversions

vishal423 commented:
@c0d33ngr, could you please resolve conflicts to allow GitHub Actions to trigger. Further, I suggest enabling the tests in one of the GitHub actions so that we can verify the tests execution result.

c0d33ngr commented:
Alright @vishal423 I'll address both issues. Thank you for your patience.

c0d33ngr commented:
async function loginByApi(request, username, password) {

    const loginResponse = await request.get('api/authenticate', {
        form: true,
        body: {
            username,
            password,
            'remember-me': true,
        }
    });
    expect(loginResponse.status()).toEqual(200);
    const csrfCookie = loginResponse.headers()['set-cookie'].split(';')[0].split('=')[1];
    console.log(username);
    console.log(password);

    const loginResponseWithCookie = await request.post('api/authentication', {
        form: true,
        headers: {
            'X-XSRF-TOKEN': csrfCookie,
        },
        body: {
            username,
            password,
            'remember-me': true,
        },
    });

    console.log(loginResponseWithCookie);
    expect(loginResponseWithCookie.status()).toEqual(200);
    const accountResponse = await request.get('api/account');
    expect(accountResponse.status()).toEqual(200);
}

Hi @vishal423 this snippet of code is a rough draft of the loginByApi function with csrfcookie. The ``expect(loginResponseWithCookie.status()).toEqual(200);`` throws 403 status code. The username and password both are correct credentials with cookie too

You mind giving any suggestions on how to go about it?

vishal423 commented:
>  The expect(loginResponseWithCookie.status()).toEqual(200); throws 403 status code. 

403 points to CSRF validation failure. I suspect the cookie that you pass along authenticate request is not valid. If you take a look at cypress equivalent commands, we read the csrf value from cookie name `XSRF-TOKEN`
>return cy.getCookie('XSRF-TOKEN').should('have.property', 'value')

As a side note, In the initial request (GET), you don't need to pass form/parameters as that's used to retrieve the authenticated user info/csrf cookie.

c0d33ngr committed:
Refactor code to follow best practices

c0d33ngr committed:
Add playwright workflow

c0d33ngr committed:
To fix merge conflict

c0d33ngr committed:
Refactor more code to follow best practices

c0d33ngr committed:
Fix syntax errors

c0d33ngr committed:
Add login with api and fix some errors

c0d33ngr committed:
Add api POM and update test cases

c0d33ngr committed:
Fix some bugs

c0d33ngr committed:
Add conditional statements to some test cases

c0d33ngr committed:
Add more conditional statements to test cases

vishal423 commented:
@c0d33ngr, do you plan to continue on this PR?

c0d33ngr commented:
> @c0d33ngr, do you plan to continue on this PR?

Yes I intend on doing so.

vishal423 commented:
@c0d33ngr, Thanks. Lets scope test scenarios excluding `entity` in this PR. Let me know if you also want to scope authentication type (between `JWT`, `Session`, `OAuth2`) in this initial PR. 

Meantime, I will start on moving `cypress` under a specific option/flag so that we can give an option to choose between `cypress` and `playwright` (later once becomes stable).

c0d33ngr commented:
> @c0d33ngr, Thanks. Lets scope test scenarios excluding `entity` in this PR. Let me know if you also want to scope authentication type (between `JWT`, `Session`, `OAuth2`) in this initial PR.
> 
> Meantime, I will start on moving `cypress` under a specific option/flag so that we can give an option to choose between `cypress` and `playwright` (later once becomes stable).

Okay. `entity` could be in another PR same as authentication scope. The only issue I'm having in this PR is fixing the merge conflict from my end

vishal423 commented:
ok. I will take a close look over the upcoming weekend.

Couple of initial observations:
- All templates files should be moved to `generators/client/templates` instead of `generators/client/templates/svelte`
- I couldn't understand the use of `generators/client/templates/svelte/playwright-e2e/jdl-config.js` contents. We use [`ejs`](https://ejs.co/) templates to access user selected or default options during code generation and the variables inside your file should be accessible to you as those are in other template files (cypress or others).
e.g. Within ejs script, the variable variable `databaseTypeSql` with appropriate value should be available and you don't need to add new logic to create or populate this to use in your scripts.
- We also need update to [package.json](https://github.com/jhipster/generator-jhipster-svelte/blob/main/generators/client/templates/package-template.json.ejs) scripts to run playwright tests. Currently we use following for cypress, so, you can create equivalent for playwright:
```
	"e2e": "npm run e2e:ci -- --headed",
	"e2e:ci": "cypress run --browser chrome",
        "cy:open": "cypress open --e2e --browser chrome",
```

c0d33ngr commented:
> @c0d33ngr, Thanks. Lets scope test scenarios excluding `entity` in this PR. Let me know if you also want to scope authentication type (between `JWT`, `Session`, `OAuth2`) in this initial PR.
> 
> Meantime, I will start on moving `cypress` under a specific option/flag so that we can give an option to choose between `cypress` and `playwright` (later once becomes stable).

Okay. `entity` could be in another PR same as authentication scope. The only issue I'm having in this PR is fixing the merge conflict from my end

> ok. I will take a close look over the upcoming weekend.
> 
> Couple of initial observations:
> 
>     * All templates files should be moved to `generators/client/templates` instead of `generators/client/templates/svelte`
> 
>     * I couldn't understand the use of `generators/client/templates/svelte/playwright-e2e/jdl-config.js` contents. We use [`ejs`](https://ejs.co/) templates to access user selected or default options during code generation and the variables inside your file should be accessible to you as those are in other template files (cypress or others).
>       e.g. Within ejs script, the variable variable `databaseTypeSql` with appropriate value should be available and you don't need to add new logic to create or populate this to use in your scripts.
> 
>     * We also need update to [package.json](https://github.com/jhipster/generator-jhipster-svelte/blob/main/generators/client/templates/package-template.json.ejs) scripts to run playwright tests. Currently we use following for cypress, so, you can create equivalent for playwright:
> 
> 
> ```
> 	"e2e": "npm run e2e:ci -- --headed",
> 	"e2e:ci": "cypress run --browser chrome",
>         "cy:open": "cypress open --e2e --browser chrome",
> ```

Thank you for the observations.

* About the ```jdl-config.js``` file, it was created so I could use the predefined user selected or default options to substitue the [ejs](https://ejs.co) but I would have to look through it to use it in playwright with little help from you if the need arises
* About the [package.json](https://github.com/jhipster/generator-jhipster-svelte/blob/main/generators/client/templates/package-template.json.ejs) file you pointed me to, I'll create equivalent playwright commands in this PR

vishal423 commented:
@c0d33ngr, Are you still interested to take it to completion? Would you like to collaborate with @Pixel998 as he also opened PR for same work https://github.com/jhipster/generator-jhipster-svelte/pull/2432

c0d33ngr commented:
Closed this in favour of #2432
